### PR TITLE
Improve experience for overnight `okteto up` sessions, and syntching timeouts

### PR DESCRIPTION
**Pull request by andreafalzetti**
# Proposed changes

This change aims to improve the retry mechanism currently affecting some users. The problem is summarised as follows:

When leaving the `okteto up` running overnight, the OS will pause the okteto CLI and other processes, causing a disconnection from the pod. When this happens, syncthing remote is not reachable, because the k8s port forwarding is also unavailable, therefore users may experience a full disconnection when starting their working day in the morning because we do not consider transient the syncthing timeout error.

After discussing different approaches, we agreed that, making permanently transient the timeout error would be incorrect, because syntching can fail to start due to configuration issues (ie. permissions), but, if the `okteto up` sequence was successfully completed once, we can exclude that syntching has configuration issues to start again, so this proposed change, is making that specific error to be considered transient depening on the previous successful execution. 

Note: there is a scenario that we should consider in the future, to listen for deployments, and make sure that the dev container has not been updated outside of the `okteto up` sequence. It could be that a user successfully runs `okteto up`, then break syncthing with another command, then we would consider the syncthing's timeout error transient when it's not. I think it's quite an edge case that can be considered as follow-up from this PR.

## How to validate

To test this change, it's quite tricky because we need to make `okteto up` suceed once, then break syncthing and trigger a reconnection, making only the remote ping to fail, and verify that it's considered transient.

The way I tested is by making the ping path dynamic, so that based on time of when I was starting `okteto up`, and simulating a disconnection by killing the pod, when the CLI tries to reconnect, for remote, the ping path is a 404, so after a few retries, it would trigger the timeout.

1. Modify `syncthing.go` to use a variable for the ping path `rest/system/ping`
1. In the `Ping` func, based on time, change this value for remote:
```go
if !local {
	// change the following time based on the execution of your debug
	if time.Now().Hour() >= 16 && time.Now().Minute() >= 19 && time.Now().Minute() <= 24 {
		s.PingPath = "/wrong/path"
		// note: s.PingPath does not exist, and should be added to the Syncthing struct
	}
}
```

## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines

### CODE REVIEW

File: cmd/up/errors.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+// isTransient is an extension of the oktetoErrors.IsTransient, this variant is used to add transient errors dynamically
+// based on the state of the upContext, in particular, if the up session was successfully started once before any retry
+func (up *upContext) isTransient(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if up.success {
+		if strings.Contains(err.Error(), "syncthing local=false didn't respond after") {
```
Comment (pchico83) [inline cmd/up/errors.go:30]: there are more errors that might appear here. what about retrying for X times if `up.success` for any kind of error?

File: cmd/up/up.go [hunk 2]
```diff
@@ -253,19 +253,21 @@ func Up(at analyticsTrackerInterface, ioCtrl *io.IOController, k8sLogger *io.K8s
...
+				StartTime:                    time.Now(),
+				Registry:                     registry.NewOktetoRegistry(okteto.Config{}),
+				Options:                      upOptions,
+				Fs:                           afero.NewOsFs(),
+				analyticsTracker:             at,
+				analyticsMeta:                upMeta,
+				K8sClientProvider:            okteto.NewK8sClientProviderWithLogger(k8sLogger),
+				tokenUpdater:                 newTokenUpdaterController(),
+				builder:                      buildv2.NewBuilderFromScratch(at, ioCtrl),
+				unhandledTransientMaxRetries: 100,
```
Comment (andreafalzetti) [inline cmd/up/up.go:269]: this `100` is very arbitrary, I think that `100` could be a good value to start testing with because we need a value that is a bit high, in case the retries need to run for hours overnight.

File: cmd/up/errors.go (hunk 1 above)
Comment (andreafalzetti) [reply cmd/up/errors.go:31]: @pchico83 thanks for the feedback, I've updated the PR

File: cmd/up/errors.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+		return false
+	}
+
+	isTransientErr := oktetoErrors.IsTransient(err)
+
+	if up.success {
+		if strings.Contains(err.Error(), "syncthing local=false didn't respond after") {
+			return true
+		}
+		if !isTransientErr && up.unhandledTransientRetryCount < up.unhandledTransientMaxRetries {
```
Comment (jLopezbarb) [inline cmd/up/errors.go:35]: Can we add a log here in order to track while debugging?

File: cmd/up/up.go (hunk 2 above)
Comment (jLopezbarb) [reply cmd/up/up.go:269]: How are we going to know if it's high or low? do we have any metric that we could add in order to know what's the maximum value?

**Review by jLopezbarb — APPROVED** [review]
Minor comments. LGTM

File: cmd/up/errors.go (hunk 3 above)
Comment (andreafalzetti) [reply cmd/up/errors.go:35]: I was planning to add a log, but then realized that we can infer by looking at the logs, if a non transient error is triggering a new sequence, it means it was treated as transient. But now, explaining this, it makes me realize that a log is just easier :) Will add

Comment (andreafalzetti) [reply cmd/up/errors.go:36]: log added in c488d90b

cc @jLopezbarb feel free to suggestion a different log message

File: cmd/up/up.go (hunk 2 above)
Comment (andreafalzetti) [reply cmd/up/up.go:269]: @jLopezbarb with the new debug log added with c488d90b, we should be able to see the total max retries and current count, when looking at the logs. Do you have more ideas?

**Review by ifbyol — APPROVED** [review]
